saleor/checkout/steps.py

```diff
@@ -16,48 +16,118 @@
+            self.email = request.user.email
+            queryset = request.user.addresses.all()
+            addresses = list(queryset)
+            self.addresses = addresses
+
+            if not self.get_same_own_address(self.address):
+                address = self.address
+            else:
+                address = None
+
+            new_address_form = AddressForm(request.POST or None,
+                                           prefix=self.step_name,
+                                           instance=address)
+
+            self.forms['addresses_form'] = UserAddressesForm(
+                data=request.POST or None, queryset=queryset,
+                prefix=self.step_name)
+
+            if not new_address_form.is_bound:
+                selected_address = self.get_same_own_address(self.address)
+                if selected_address:
+                    selected_address.is_selected = True
+                    self.is_new_address = False
+                elif not address_data:
+                    default_address = request.user.default_shipping_address
+                    if default_address:
+                        address = self.get_same_own_address(default_address)
+                        if address:
+                            address.is_selected = True
+                            self.is_new_address = False
+                    elif addresses:
+                        addresses[0].is_selected = True
+                        self.is_new_address = False
         else:
-            addresses = []
-        self.existing_selected = existing_selected
-        self.forms = {'address': address_form}
-        self.addresses = addresses
+            new_address_form = AddressForm(request.POST or None,
+                                           prefix=self.step_name,
+                                           instance=self.address)
+            email = storage.get('email', '')
+            self.email = email
+            self.forms['email'] = AnonymousEmailForm(request.POST or None,
+                                                     initial={'email': email})
+        self.forms['new_address'] = new_address_form
+
+    def get_same_own_address(self, address):
+        for own_address in self.addresses:
+            if Address.objects.are_identical(address, own_address):
+                return own_address
+
+    def process(self, extra_context=None):
+        context = dict(extra_context or {})
+        context['addresses'] = self.addresses
+        context['new_address'] = self.is_new_address
+        context['button_label'] = _('Ship to this address')
+        return super(BaseCheckoutStep, self).process(extra_context=context)
 
     def forms_are_valid(self):
-        address_form = self.forms['address']
-        return address_form.is_valid()
+        address_is_valid = False
+        if self.addresses:
+            addresses_form = self.forms['addresses_form']
+            if addresses_form.is_valid():
```

Comment:
Can't we always depend on this form being there? Even for anonymous users/those without existing addresses we can always render the only choice as a hidden field. Currently it's not obvious how the logic flows through this method.